api/apps/api/src/modules/features/domain/feature-specification.spec.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  FeatureSpecification,
+  SpecificationStatus,
+} from './feature-specification';
+import { FeatureSpecificationId } from './feature-specification.id';
+import { FeatureSpecificationRevision } from './feature-specification-revision';
+
+describe(`when specification was already created`, () => {
+  let spec: FeatureSpecification;
+  beforeEach(() => {
+    spec = new FeatureSpecification(
+      new FeatureSpecificationId(`spec-id`),
+      SpecificationStatus.Created,
+      [],
+      new FeatureSpecificationRevision(5),
+    );
+  });
+
+  describe(`when submitting again`, () => {
+    beforeEach(() => {
+      spec.submit();
+    });
+
+    it(`should emit compute event`, () => {
+      expect(spec.getUncommittedEvents()).toMatchInlineSnapshot(`
+        Array [
+          ComputeFeatureSpecificationEvent {
+            "featureSpecificationId": "spec-id",
+            "revision": 6,
+          },
+        ]
+      `);
+    });
+
+    it(`should change state`, () => {
+      expect(spec.getStatus()).toEqual(SpecificationStatus.Created);
+    });
+    it(`should set revision to 6`, () => {
+      expect(spec.getRevision()).toEqual(6);
+    });
+  });
+
+  describe(`when changing spec`, () => {
+    beforeEach(() => {
+      spec.change({
+        features: [],
+      });
+    });
+
+    it(`should not emit compute event`, () => {
+      expect(spec.getUncommittedEvents()).toEqual([]);
+    });
+    it(`should put into draft state`, () => {
+      expect(spec.getStatus()).toEqual(SpecificationStatus.Draft);
+    });
```

Comment:
I would say that it may be possible for a user to change a specification and submit it for processing at the same time (hence `SpecificationStatus.Created`).

Reply:
@hotzevzl I suppose this case is covered in tests.

```
when creating brand new specification > when submitting
```

so, from system/product perspective, its creating "draft" on which it will be immediately `submit`ted.

just to distinguish it, we may put some static `fresh(...payload)` method which will assign default values (or actually, it can be a domain object factory with such method)

Reply:
hence, this case would be handled by the application as:
```
const spec = SpecificationFactory.fresh(...payload) // including validation
spec.submit()
```

Reply:
yes, but in practice we ask for all the properties to be submitted (unless they are optional, of course), and we don't need to compare with what we already have (if anything) - or do you mean to compare the stored specification with the one in a new request? If so I would just hash a normalized serialization - if hash doesn't match => new revision (could even match a previous one, we shouldn't care, I think).

`when changing spec > it should put into draft state` should not necessarily be the case - but I may be misreading the intention behind this test.

I would treat each request as a full replacement of any previous specification, and except for the hashing bit if we want to keep track of revisions to see if what is calculated matches what the user can `GET`, I don't think we need to have default values - the optional bits are really optional (a feature set can be empty, mostly by user mistake, and we will not have `prop` and `fpf` until the second step of the process in the UI, but other than that, we should always receive a `status` and enforce that is present and is one of the allowed enum values.

Then we can use per-geoprocessing operation hashing to check if we need to *recompute* an operation, but that's a separate story.

Reply:
@hotzevzl I am a bit lost and confused - may be that we are misreading the purpose of this domain class.

It totally abstracts from what is within specification, it "only" handles draft/created states transitions according to specification and what happens with the revision once one of the actions is done. It neither knows the details, nor the user context, nor the recompute needs (as you mentioned, geo will take care of that).

Especially two cases:
```
// submit -> submit again ; which is actually creating a new versions
if the specification is submitted again with status `created`, resulting
  features will be computed again, any existing features linked to the scenario
  will be unlinked from it, and the newly calculated features will be linked to
  it
```

```
// it isn't the case from user perspective, as he wants to modify & submit but that actually creates a new version (think of Conflucence page modification)
"when changing spec > it should put into draft state" should not necessarily be the case - but I may be misreading the intention behind this test.
```

Reply:
yes, it's fine that this is only about the state transitions.

the Confluence example helps, thanks. I was thinking about how our own frontend is using these interim `PUT`s, but yes, if a user were to only update "something else except the status" we should indeed assume that the new copy is a draft. If we enforce `status` to be present in the payload (which we do, IIRC) this may be immaterial, but ok in any case.